### PR TITLE
`Expr::evaluate` method for `CatOp` and `SliceOp`

### PR DESCRIPTION
**Pull request by Priya2698**
Implements the `Expr::evaluate` method for `CatOp` and `SliceOp`. This is a part of resolving issue #670.

### CODE REVIEW

**Comment by Priya2698** [comment]
This should be rebased after merging PR #881.

**Comment by Priya2698** [comment]
!build

**Comment by Priya2698** [comment]
!build

File: csrc/expr_evaluator.cpp
Comment (jacobhinkle) [inline csrc/expr_evaluator.cpp:204]: This behaves a bit different from the non-const evaluate, since it is restricted to known values only so it is "shallow" like `getValue`. Could we instead consider making `getValue` const (which it should be) and further just exposing that as a public method? That way it is clear that `getValue` does a shallow but const lookup and `evaluate` does a deep but non-const one. Alternatively, we could implement evaluate this const version recursively just like the non-const one, but just avoid the step where we set values in `known_values_`.

File: test/test_resize.cpp
Comment (jacobhinkle) [inline test/test_resize.cpp:1805]: I have a question for @zasdfgbnm and maybe @naoyam here. I get that using the evaluator for the ATen ground truth is useful in case we don't want to provide the manually-defined reference. However, if we're testing a chunk of code that modifies the fusion graph, we might want to keep the ATen reference since the automatic method can only check that the output is consistent with the graph itself. I am thinking mostly of the resize and dynamic shape tests where there is a danger of modifying the graph in a way that subtly changes the output even though it's self-consistent. For example if we replaced a pad with `set` in some op, the automatic tests might not catch the bug. Since we already have working references in existing tests, maybe we should just leave them, but feel free to omit for new tests? What do you guys think?

Comment (naoyam) [reply test/test_resize.cpp:1805]: That's a very good point. Can we use the original fusion since it should not be modified?

Comment (jacobhinkle) [reply test/test_resize.cpp:1805]: Actually yes this should use the unmodified fusion `executor_cache.fusion()`. If we instead grabbed the last compiled runtime from the cache and used its complete fusion as the reference, we might get something different. I guess I hadn't considered that we're moving unique pointers around, so there's no real danger of unwittingly passing the modified fusion here.

**Review by zasdfgbnm — APPROVED** [review]
My reading for the conversation https://github.com/NVIDIA/Fuser/pull/882#discussion_r1336273119
Is that Priya is already doing the correct thing.
The rest of this PR LGTM as well. Please dismiss my approval if I am reading the conversation wrong.

**Comment by jacobhinkle** [comment]
> My reading for the conversation [#882 (comment)](https://github.com/NVIDIA/Fuser/pull/882#discussion_r1336273119) Is that Priya is already doing the correct thing. The rest of this PR LGTM as well. Please dismiss my approval if I am reading the conversation wrong.

Yes definitely! Sorry I should've been more clear: my fears were allayed by @naoyam. Merge at will.

**Pull request merged**
